code/modules/library/library_admin.dm

```diff
@@ -0,0 +1,212 @@
+			switch(id)
+				if("specify_ssid_delete")
+					ui_modal_input(src, id, "Please input a book SSID:", null, arguments)
+				if("specify_ckey_search")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				if("specify_ckey_delete")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				else
+					return FALSE
+		if(UI_MODAL_ANSWER)
+			var/answer = params["answer"]
+			switch(id)
+				if("specify_ssid_delete")
+					if(!answer)
+						return
+					var/confirm = alert("You are about to delete book [text2num(answer)]", "Confirm Deletion", "Yes", "No")
+					if(!confirm || confirm == "No")
+						return
+					if(GLOB.library_catalog.removeBookByID(text2num(answer)))
+						log_admin("[key_name(usr)] has deleted the book [text2num(answer)].")
+						message_admins("[key_name_admin(usr)] has deleted the book [text2num(answer)].")
+				if("specify_ckey_search")
+					if(!answer)
+						return
+					var/datum/library_user_data/search_terms = new()
+					search_terms.search_ckey = answer
+					selected_ckey = answer
+					cached_books = list()
+					for(var/datum/cachedbook/CB in GLOB.library_catalog.getBooksByRange(1, 10, search_terms))
+						var/list/book_data = list(
+							"id" = CB.id,
+							"title" = CB.title,
+							"author" = CB.author,
+							"rating" = CB.rating,
+							"summary" = CB.summary,
+							"ckey" = CB.ckey,
+							"reports" = CB.reports,
+						)
+						cached_books += list(book_data)
+					page_state = LIBRARY_MENU_CKEY
+				if("specify_ckey_delete")
+					if(!answer)
+						return
+					var/confirm //We want to be absolutely certain an admin wants to do this
+					confirm = alert("You are about to mass delete potentially up to 10 books", "Confirm Deletion", "Yes", "No")
+					if(!confirm || confirm == "No")
+						return
+					if(GLOB.library_catalog.removeBooksByCkey(answer))
+						log_admin("[key_name(usr)] has deleted all books uploaded by [answer].")
+						message_admins("[key_name_admin(usr)] has deleted all books uploaded by [answer].")
+				else
+					return FALSE
+		else
+			return FALSE
+
+/datum/ui_module/library_manager/proc/view_book_by_id(bookid)
+	if(view_book?.id != bookid) //lets not make more queries than neccesary
+		view_book = GLOB.library_catalog.getBookByID(bookid)
+	view_book_page = 0
+	view_book()
+
+/*
+* #View Book
+*
+* Internal proc for viewing library books as an admin. This absolutely must stay as BrowserUI even though the rest of
+* the library manager is TGUI. This is because of TGUI sanitization issues.
+*/
+/datum/ui_module/library_manager/proc/view_book(mob/user)
+	if(!view_book || !length(view_book.content))
+		return
+
+	var/dat = " "
+	//First, we're going to choose/generate our header buttons for switching pages and store it in var/dat
+	if(length(view_book.content)) //No need to have page switching buttons if there's no pages
+		if(view_book_page == 0)
+			dat+= "<DIV STYLE='float:left; text-align:left; width:49.9%'></DIV>"
```

Comment:
Dupe logic from earlier init? Best to reuse some code

Reply:
yes, I'll try and find a solution to this. datum perhaps?

Reply:
I will do my best to find an alternative to datumizing this. Although we may need to make do right now with a bit of duplicate code